File: Insert-Greatest-Common-Divisors-in-Linked-List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
    int GCD(int a,int b){
        if(b==0){
            return a;
        }
        return GCD(b,a%b);
    }
public:
    ListNode* insertGreatestCommonDivisors(ListNode* head) {
        if(head->next == nullptr){
            return head;
        }
        ListNode* prev = head;
        ListNode* curr = head->next;
        while(prev && curr){
            int output = GCD(max(prev->val,curr->val),min(prev->val,curr->val));

            ListNode* new_node = new ListNode(output);
            prev->next= new_node;
            new_node ->next = curr; 

            prev = curr;
            curr = curr->next;
        }
    return head ; 
    }
};

File: Insert-Greatest-Common-Divisors-in-Linked-List.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    private int GCD(int a, int b) {
        return (b == 0) ? a : GCD(b, a % b);
    }
    public ListNode InsertGreatestCommonDivisors(ListNode head) {
        if (head.next == null) {
            return head;
        }

        ListNode prev = head;
        ListNode curr = head.next;  

        while (curr != null) {
            int gcdValue = GCD(Math.Max(prev.val, curr.val), Math.Min(prev.val, curr.val));


            ListNode newNode = new ListNode(gcdValue);
            prev.next = newNode;
            newNode.next = curr;

            prev = curr;
            curr = curr.next;
        }

        return head;
    }
}